polyhost/services/updater.py

```
@@ -0,0 +1,226 @@
+
+def _safe_extract(tar: tarfile.TarFile, dest: Path) -> None:
+    """Extract a tarball, refusing any member that resolves outside `dest`."""
+    if sys.version_info >= (3, 12):
+        tar.extractall(path=dest, filter="data")
+        return
+    dest_resolved = dest.resolve()
+    for member in tar.getmembers():
+        target = (dest / member.name).resolve()
+        if dest_resolved != target and dest_resolved not in target.parents:
+            raise RuntimeError(f"Refusing unsafe tar member: {member.name}")
+    tar.extractall(path=dest)
```

Ruff (0.15.15)

[error] Uses of `tarfile.extractall()`

(S202)